gitlab/resource_gitlab_project_access_token_test.go

```diff
@@ -0,0 +1,248 @@
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabProjectAccessTokenExists("gitlab_project_access_token.bar", &pat),
+					testAccCheckGitlabProjectAccessTokenAttributes(&pat, &testAccGitlabProjectAccessTokenExpectedAttributes{
+						name:      "my project token",
+						scopes:    map[string]bool{"read_repository": true, "api": true, "write_repository": true, "read_api": true},
+						expiresAt: "2022-04-01",
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabProjectAccessTokenExists(n string, pat *testAccGitlabProjectAccessTokenWrapper) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", n)
+		}
+
+		patID := rs.Primary.ID
+		repoName := rs.Primary.Attributes["project"]
+		if repoName == "" {
+			return fmt.Errorf("No project ID is set")
+		}
+		conn := testAccProvider.Meta().(*gitlab.Client)
+
+		tokens, _, err := conn.ProjectAccessTokens.ListProjectAccessTokens(repoName, nil)
+		if err != nil {
+			return err
+		}
+
+		for _, token := range tokens {
+			if strconv.Itoa(token.ID) == patID {
+				pat.pat = token
+				pat.project = repoName
+				pat.token = rs.Primary.Attributes["token"]
+				return nil
+			}
+		}
+		return fmt.Errorf("Project Acces Token does not exist")
+	}
+}
+
+type testAccGitlabProjectAccessTokenExpectedAttributes struct {
+	name      string
+	scopes    map[string]bool
+	expiresAt string
+}
+
+type testAccGitlabProjectAccessTokenWrapper struct {
+	pat     *gitlab.ProjectAccessToken
+	project string
+	token   string
+}
+
+func testAccCheckGitlabProjectAccessTokenAttributes(patWrap *testAccGitlabProjectAccessTokenWrapper, want *testAccGitlabProjectAccessTokenExpectedAttributes) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		pat := patWrap.pat
+		if pat.Name != want.name {
+			return fmt.Errorf("got Name %q; want %q", pat.Name, want.name)
+		}
+
+		if pat.ExpiresAt.String() != want.expiresAt {
+			return fmt.Errorf("got ExpiresAt %q; want %q", pat.ExpiresAt.String(), want.expiresAt)
+		}
+
+		for _, scope := range pat.Scopes {
+			if !want.scopes[scope] {
+				return fmt.Errorf("got a not wanted Scope %q, received %v", scope, pat.Scopes)
+			}
+			want.scopes[scope] = false
+		}
+		for k, v := range want.scopes {
+			if v {
+				return fmt.Errorf("not got a wanted Scope %q, received %v", k, pat.Scopes)
+			}
+		}
+
+		git, err := gitlab.NewClient(patWrap.token, gitlab.WithBaseURL((testAccProvider.Meta().(*gitlab.Client)).BaseURL().String()))
+		if err != nil {
+			return fmt.Errorf("Cannot use the token to instantiate a new client %s", err)
+		}
+		_, _, err = git.ProjectMembers.ListAllProjectMembers(patWrap.project, nil)
+		if err != nil {
+			return fmt.Errorf("Cannot use the token to perform an API call %s", err)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGitlabProjectAccessTokenDestroy(s *terraform.State) error {
```

Comment:
You don't have anywhere in your tests where you test the behavior of your destroy function, so it's possible that tokens are not being deleted from GitLab. The `testAccCheckGitlabProjectAccessTokenDestroy` function is meant to validate that there are no more `gitlab_project_access_token`s for the project, not that the project is deleted. Depending on how comfortable you are with the test framework, it would be great if you could improve this. Here's an example where this is done right:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/446a7d13f859235caccc58ab8a0129d89a61c04d/gitlab/resource_gitlab_project_approval_rule_test.go#L166-L174

Reply:
Implemented testing of the destroy function. Had to add a sleep because GitLab immediately returns from the destroy REST call but then deletes the token lazely. So we have to wait some time before checking if the token still exists.